core/common/src/serializers/DateTimeUnitSerializers.kt

```diff
@@ -21,8 +21,10 @@ import kotlin.reflect.KClass
  */
 public object TimeBasedDateTimeUnitSerializer: KSerializer<DateTimeUnit.TimeBased> {
 
-    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("TimeBased") {
-        element<Long>("nanoseconds")
+    override val descriptor: SerialDescriptor by lazy {
```

Comment:
This bug observable only for wasm target (which seems is more strict for nullability in this case). Could you please to suggest how to file such bug and where.

Reply:
I don't know how nullability comes into play here: these are all not null and don't even depend on one another.

I'd file this under https://kotl.in/issue, specifying that the behavior is different on wasm and crashes the code that works on all other platforms.

Reply:
The initialisation sequence reads `null` value from `not-null` uninitialised property. Here, I put the code into lazy block to prevent reading uninitialised property. The bug exists at all platforms but seems it is not observable.

Reply:
I feel we're speaking past each other.

> The bug exists at all platforms but seems it is observable only on wasm.

The way I'm reading this, if this bug *never* surfaces on other platforms, then we can't call it a bug. Kotlin/Wasm behaves differently, making code that always works correctly on other platforms buggy. This is a problem with Kotlin/Wasm and needs to be reported.

Or did you mean that you *can* cause this code to misbehave on other platforms somehow, we just never encountered it, and there actually is a bug in the library code that's fixed by `lazy`? In what circumstances can it happen?

Either way, could you elaborate on what the bug is? At which point is a `null` value read from a non-`null` property?

Reply:
I am not really sure that it is worth to discuss this. 
Here: https://github.com/Kotlin/kotlinx-datetime/blob/master/core/common/src/serializers/DateTimeUnitSerializers.kt#L192
I see implementation of internal class `AbstractPolymorphicSerializer` which is not supposed to implement by user (with overriding internal members with a suppression and using internal API. So, the hacky way of implementing serialisation already here and I do not see how I should file a bug for a such code.
At least I do not have expertise for serialisation guts.
Anyway, if we look into `AbstractPolymorphicSerializer` implementations inside `serialisation` library, there is always usages of `lazy` for this class implementors. I believe it worth to ask about it for somebody who works with serialisation implementation. As for me, I would remove these `lazy` from `KSerializer` implementations and put it into `AbstractPolymorphicSerializer` implementations to stick with `serialisation` library internals.

Reply:
> I do not see how I should file a bug for a such code

Something like this: "There's a bug in the Kotlin compiler for Wasm. Here's some code that throws a `NullPointerException` on Wasm but not on any other platform." If you can think of a minimal example (that ideally fits in a single file) where this happens without any library dependencies, it would also help the people writing the compiler. If not, it's usually enough to send them a link to any project where this reproduces, like "https://github.com/Kotlin/kotlinx-datetime/blob/e41ad76db5e52d2d2e8997f82a5704b21bdba9b7/core/common/src/serializers/DateTimeUnitSerializers.kt here, removing `lazy` on line 162 causes crashes"

Reply:
I do not think this is Kotlin/Wasm compiler bug.

Reply:
Ok, where is this difference coming from then? Is the serialization library using `expect`/`actual` in this part to have different code on different platforms and uses a different implementation from everything else on Kotlin/Wasm?

Reply:
The fact that at all backends except wasm the behaviour is different does not mean that this is a kotlin/wasm bug, especially for the code that uses frontend suppressors.

Reply:
I found the difference of implementation for associated objects in the `wasm` that triggered the problem.
It was a difference of associated objects initialization that was not specified and tested.
In a matter of same behavior for different targets I have filed a [bug](//https://youtrack.jetbrains.com/issue/KT-63939) for kotlin/wasm stdlib.
As far, to workaround this difference, I put `lazy` for all serialiser constructor initialisers in `DateTime` with a comment.

Reply:
Good job figuring this one out!